PEMStoreSSH/CertificateFormats/PKCS12Handler.cs

```diff
@@ -64,16 +69,44 @@ public X509Certificate2Collection RetrieveCertificates(byte[] binaryCertificates
         public List<SSHFileInfo> CreateCertificatePacket(string certToAdd, string alias, string pfxPassword, string storePassword, bool hasSeparatePrivateKey)
         {
             List<SSHFileInfo> fileInfo = new List<SSHFileInfo>();
-            Pkcs12Store store;
 
+            X509Certificate2 x509Cert = new X509Certificate2(Convert.FromBase64String(certToAdd), pfxPassword);
+            string aliasToUse = string.IsNullOrEmpty(alias) ? x509Cert.Thumbprint : alias;
+
+            Pkcs12Store tempStore = new Pkcs12Store();
             using (MemoryStream inStream = new MemoryStream(Convert.FromBase64String(certToAdd)))
             {
-                store = new Pkcs12Store(inStream, pfxPassword.ToCharArray());
+                if (string.IsNullOrEmpty(pfxPassword))
+                {
+                    CertificateConverter converter = CertificateConverterFactory.FromDER(Encoding.ASCII.GetBytes(certToAdd));
+                    Org.BouncyCastle.X509.X509Certificate bcCert = converter.ToBouncyCastleCertificate();
+                    X509CertificateEntry entry = new X509CertificateEntry(bcCert);
+                    tempStore.SetCertificateEntry(aliasToUse, entry);
+                }
+                else
+                {
+                    tempStore = new Pkcs12Store(inStream, pfxPassword.ToCharArray());
+                }
+            }
+
+            string tempAlias = string.Empty;
+            foreach (string name in tempStore.Aliases)
+            {
+                tempAlias = name;
+                break;
+            }
+
+            Pkcs12Store store = new Pkcs12Store();
+
+            store.SetCertificateEntry(aliasToUse, tempStore.GetCertificate(tempAlias));
+            if (!string.IsNullOrEmpty(pfxPassword))
+            {
+                store.SetKeyEntry(aliasToUse, tempStore.GetKey(tempAlias), tempStore.GetCertificateChain(tempAlias));
             }
 
             using (MemoryStream outStream = new MemoryStream())
             {
-                store.Save(outStream, string.IsNullOrEmpty(storePassword) ? pfxPassword.ToCharArray() : storePassword.ToCharArray(), new Org.BouncyCastle.Security.SecureRandom());
+                store.Save(outStream, string.IsNullOrEmpty(storePassword) ? new char[0] : storePassword.ToCharArray(), new Org.BouncyCastle.Security.SecureRandom());

```

Comment:
Is saving the store without a password still secured somehow? By `SetKeyEntry` above?

Reply:
Matt, so if there is no pfx password, there is no private key, and the file does not need to be secured.